src/server.c

```diff
@@ -3558,7 +3586,10 @@ void call(client *c, int flags) {
 
         /* Call alsoPropagate() only if at least one of AOF / replication
          * propagation is needed. */
-        if (propagate_flags != PROPAGATE_NONE) alsoPropagate(c->db->id, c->argv, c->argc, propagate_flags);
+        if (propagate_flags != PROPAGATE_NONE && shouldPropagate(propagate_flags)) {
+            alsoPropagate(c->db->id, c->argv, c->argc, propagate_flags);
+            c->flags |= CLIENT_PROPAGATING;
```

Comment:
Strictly speaking, in the case where `appendfsync` is set to `always`, using synchronous I/O operations can achieve the semantics of **read committed** isolation. Because it is only after `flushAppendOnlyFile` completes both the **write** and **fsync** operations that the function `handleClientsWithPendingWritesUsingThreads` is called to return results to the client. Therefore, it can be ensured that, within this event loop, both read and write commands will only receive results from the client after AOF has completed the persistent flushing to disk.

Now, using `io_uring` and marking only write operations with the `CLIENT_PROPAGATING` flag implies that the read commands within this event loop—which may have already read data not yet persisted—will send results to the client before waiting for the `io_uring` task to complete. This effectively degrades the isolation level to **read uncommitted**.

Here, I believe we should maintain the **read committed** isolation level. A simple idea is not to distinguish between read and write commands. If `io_uring` is enabled, we could consider delaying the invocation of `handleClientsWithPendingWritesUsingThreads` until after `afterSleep`, which is post `epoll_wait`, giving asynchronous I/O more time to execute and thus reducing wait time in the main thread.

Reply:
>I believe we should maintain the read committed isolation level.

Good point @soloestoy. Agreed.

>we could consider delaying the invocation of handleClientsWithPendingWritesUsingThreads until after afterSleep

I realize that it is hard to time the pace of two "tracks": the command processing and the fsync work. If we have a slower disk, I think it makes sense to delay `writeToCilent` until after `afterSleep` but if the disk is fast (some enterprise grade NVMes could complete fsync in less than 0.1 ms), adding an O(N) `epoll` call could end up nullifying the benefit of the async fsync, or even regressing the performance under the same situation, I am guessing.

Reply:
@soloestoy @PingXie Thanks for you comments and effort on this.

But I still want to argue for the patch. Please feel free to correct me if I was misunderstanding, or you can just close the topic if your core team have already made the decision. :)

For the `read committed` contract, I am wandering if we had any official announcement that we support this isolation when `AOF=always` is enabled? Because I think the `read committed isolation` is not true based on existing implementation. Maybe we should drop the group commit firstly. 

From the definition of read committed I got from wikipedia：

> Putting it in simpler words, read committed is an isolation level that guarantees that any data read is committed at the moment it is read. It simply restricts the reader from seeing any intermediate, uncommitted, 'dirty' read. It makes no promise whatsoever that if the transaction re-issues the read, it will find the same data; data is free to change after it is read.

The `read committed` we defined in Valkey is a little confused here, it depends on the moment that fsync to disk successfully rather than the moment (actually this is the real moment that a read client got value from) that data written to memory. Imagine like this, we have client A, client B and client C in one group, the sequential execution is like this:  client A `set a = 'x'`, client B `get a` then 'x' is returned, client C `set a = 'xx'`, after data committed, `a == 'xx'`, but the response data for client B 'X', is it called intermediate value or dirty read? 

IMO, when `AOF = always`, we only can guarantee that it is a more strict backup strategy than `every seconds`, to reduce the data loss from maximum 2 seconds to less, the cost is the performance, and we should avoid to give user a illusion that their data are correct. Because it still have risky that caused the data corruption, for example, both client A and client B have set commands, and they are in a same group, will call fysnc once to flush to disk, but fsync is not atomic, it will have opportunity that after client A's command flushed successfully but failed at client B's command(e.g. disk no space), then server crash. The response to user is that both client A and client B are written failed, but the fact is AOF contains client A's commands, this will cause data corruption to user. 

In conclusion, before supported the ACID, I think `read committed isolation` is not true and confused.

Reply:
A: set a 'x'
    B: get a
    C: set a 'xx

B can receive 'x', because there is no ordering between clients. If B receives 'x', then A receives +OK for the set command, so this value has been observed to be written (and overwritten by C and then fsync'ed before these clients got their replies). It's not an intermediate value or dirty read. That's what I think, but I wish to hear others reasoning too.

I didn't think about *read committed* earlier when I reviewed this PR. I don't know what we can do now, but this work about the io_uring abstractions should useful for writes using io_uring and other features. We should definitely introduce io_uring in Valkey in some ways.

* Can we use io_uring to perform async fsync for AOF = every second?
* Can we introduce a new AOF config option like AOF = asap, which means we do fsync after every write but without the *read committed* guarantee?

Reply:
>The read committed we defined in Valkey is a little confused here, it depends on the moment that fsync to disk successfully rather than the moment (actually this is the real moment that a read client got value from) that data written to memory. Imagine like this, we have client A, client B and client C in one group, the sequential execution is like this: client A set a = 'x', client B get a then 'x' is returned, client C set a = 'xx', after data committed, a == 'xx', but the response data for client B 'X', is it called intermediate value or dirty read?

I think we are actually looking at "prefix consistency" where readers can read any writes but will not see the writes *out-of-order*, even in the failure case. Here is an example how this can be violated with the current PR

```
0. a = 'x'
1. // batch begins
2. A: set a 'y'
3. B: get a
4. // batch ends
5. release B
6. fsync // and failed, then crashed the server
7. release A
```

`B` gets a value `y` that was neither ack'd back to `A` nor present after the server restarts (and restores from the AOF file)

> The response to user is that both client A and client B are written failed, but the fact is AOF contains client A's commands, this will cause data corruption to user.

Since we are talking about `ALWAYS`, if the fsync failed, there shouldn't be any response back to the user. Whether a torn fsync will lead to the AOF file being corrupted though is a separate issue.

Reply:
>Can we use io_uring to perform async fsync for AOF = every second?

Only `ALWAYS` fsync is performed synchronously. `everysecond` is already done in the background so io_uring's benefit is not as much.

>Can we introduce a new AOF config option like AOF = asap, which means we do fsync after every write but without the read committed guarantee?

I doubt this will work. The amount of fsync work in this case will likely result in the foreground `write()` call getting blocked too and slowing down the main thread. More importantly, I don't think there is a materially higher level guarantee for durability. We are looking at an RPO somewhere between 0s and 1s but as long as we don't get to 0s we still don't have the level of durability one would expect from a canonical transaction.

Reply:
I think @PingXie 's example is great, and sufficiently illustrates the issue. The essence is that when `appendfsync` is set to always, we should not only ensure that write commands are persisted, but also that the results of the write commands, as well as the results of the read commands following these modifications, are only returned to the client after AOF persistence has been completed.